### PR TITLE
Make large message handling more robust

### PR DESCRIPTION
**Pull request by ConnorMcMahon**
PR #435 attempted to fix large message inconsistencies in how it handled UTF8 vs UTF16 character counts, but it turns out that the code path that starts new orchestrations relied on the overcounting of characters that used to occur.

This PR fixes that case, and adds more testing with characters of varying UTF8 and UTF16 sizes to make sure we cover all of the edge cases.

Addresses #448

### CODE REVIEW

File: src/DurableTask.AzureStorage/Partitioning/AppLeaseManager.cs [hunk 1]
```diff
@@ -319,12 +319,8 @@ public async Task DeleteContainerAsync()
                     await this.appLeaseContainer.DeleteIfExistsAsync();
                 }
             }
-            catch (StorageException ex)
+            catch (StorageException)
```
Comment (ConnorMcMahon) [inline src/DurableTask.AzureStorage/Partitioning/AppLeaseManager.cs:322]: This should probably be covered with this PR: https://github.com/Azure/durabletask/pull/440.

File: src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs [hunk 2]
```diff
@@ -1194,7 +1196,21 @@ static string GetBlobName(DynamicTableEntity entity, string property)
...
-            string eventType = entity.Properties["EventType"].StringValue;
+            string eventType;
+            if (entity.Properties.ContainsKey("EventType"))
+            {
+                eventType = entity.Properties["EventType"].StringValue;
+            }
+            else if (property == "Input")
+            {
+                // This queue is not actually tracked as input, so just use any name.
+                eventType = "Input";
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:1207]: I'm not understanding what this comment is conveying. If I understand the code change, you're doing this because the Instances table entities don't have an `EventType` property. Is that correct?  If so, I feel like we should communicate that in the comments. Also, I'm not sure why this comment here is talking about queues, since I don't believe `AzureTableTrackingStore` knows anything about queues, nor is `DyanamicTableEntity` associated with queues. Should we instead be saying something like "Entities for the Instances table don't have an EventType property, so we are using a hardcoded value instead to represent the orchestration input.

Comment (ConnorMcMahon) [reply src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:1207]: I'm not exactly sure what I was intending to convey with this comment either, and I don't know how queue got in there at all. The reason is that the initial row does not have an eventType, but it still has an input, so we use a hardcoded value instead. I will update the comment to reflect this.

File: src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs [hunk 3]
```diff
@@ -836,6 +836,8 @@ public override async Task<PurgeHistoryResult> PurgeInstanceHistoryAsync(string
                 }
             };
 
+            await this.CompressLargeMessageAsync(entity);
```
[the remark this answers is not in the document]
Comment (ConnorMcMahon) [reply src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:839]: Before, we relied on the fact that the queue message was already serialized as a blob for these large messages, and just used the large blob name from the queue message for the input. This worked because we were counting with UTF16 instead of UTF8.

Now, it is possible that the blob message was able to fit, but it is still too large for the table, so we need to compress the message first.

[the remark this answers is not in the document]
Comment (cgillum) [reply src/DurableTask.AzureStorage/Tracking/AzureTableTrackingStore.cs:839]: I think you meant that the _queue_ message was able to fit, but otherwise, this makes sense. Thanks!

**Review by cgillum — APPROVED** [review]
LGTM! Also, thanks for taking the time to add these new test!

File: test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs [hunk 4]
```diff
@@ -1443,19 +1530,39 @@ public async Task LargeTextMessagePayloads_FetchLargeMessages_QueryState(bool en
...
+            {
+                throw new InvalidOperationException($"No characters have byte size {utf16ByteSize} for UTF16");
+            }
+            else if (utf8ByteSize < 1 || utf8ByteSize > 4)
+            {
+                throw new InvalidOperationException($"No characters have byte size {utf8ByteSize} for UTF8.");
+            }
+            else if (utf8ByteSize == 1 && utf16ByteSize == 2)
+            {
+                // Use a character set that is small for UTF16 and large for UTF8
```
Comment (ChenglongLiu) [inline test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs:1546]: Are the UTF8 and UTF16 in the wrong order?

Comment (ConnorMcMahon) [reply test/DurableTask.AzureStorage.Tests/AzureStorageScenarioTests.cs:1546]: Yeah, this comment is incorrect. This looks to be a copy paste problem. Will resolve before merging.
